### PR TITLE
Added support for the Meson Build System

### PR DESCRIPTION
**Pull request by eidheim**
Should be 100%, but some additional testing needed. Let me know if there are any issues or not.

### CODE REVIEW

**Comment by zalox** [comment]
I'll have a look, please give me some time to learn meson.

**Comment by eidheim** [comment]
Sure, and thank you:)

File: src/compile_commands.cc [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+  bool found_argument=false;
+  for(auto &parameter: parameters) {
+    if(found_argument) {
+      parameter_values.emplace_back(parameter);
+      found_argument=false;
+    }
+    else if(parameter==paramter_name)
+      found_argument=true;
+  }
+
```
Comment (zalox) [inline src/compile_commands.cc]:
```c++
std::copy_if(parameters.begin(),paramerters.end(),parameter_values.end(),[](){}));
```

File: src/compile_commands.cc [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+  }
+  
+  return parameter_values;
+}
+
+CompileCommands::CompileCommands(const boost::filesystem::path &build_path) {
+  try {
+    boost::property_tree::ptree root_pt;
+    boost::property_tree::json_parser::read_json((build_path/"compile_commands.json").string(), root_pt);
+
```
Comment (zalox) [inline src/compile_commands.cc]: Correct me if I'm wrong, but I think read_json can cause problems. I think there is a error_code version.

File: src/compile_commands.cc [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        parameters.emplace_back(parameter);
+      };
+      for(size_t c=0;c<parameters_str.size();++c) {
+        if(backslash)
+          backslash=false;
+        else if(parameters_str[c]=='\\')
+          backslash=true;
+        else if((parameters_str[c]==' ' || parameters_str[c]=='\t') && !backslash && !single_quote && !double_quote) {
+          if(parameter_start_pos!=static_cast<size_t>(-1)) {
+            add_parameter();
```
Comment (zalox) [inline src/compile_commands.cc]: std::npos?

File: src/meson.cc [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    return false;
+  if(!boost::filesystem::exists(default_build_path)) {
+    boost::system::error_code ec;
+    boost::filesystem::create_directories(default_build_path, ec);
+    if(ec) {
+      Terminal::get().print("Error: could not create "+default_build_path.string()+": "+ec.message()+"\n", true);
+      return false;
+    }
+  }
+
```
Comment (zalox) [inline src/meson.cc]: Suggestion:
```c++
if(project_path.empty() || default_build_path.empty())
    return false;
if(!boost::filesystem::exists(project_path/"meson.build"))
   return false;
```

File: src/meson.cc [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+    return false;
+  if(!boost::filesystem::exists(debug_build_path)) {
+    boost::system::error_code ec;
+    boost::filesystem::create_directories(debug_build_path, ec);
+    if(ec) {
+      Terminal::get().print("Error: could not create "+debug_build_path.string()+": "+ec.message()+"\n", true);
+      return false;
+    }
+  }
+
```
Comment (zalox) [inline src/meson.cc]: This calls for abstraction, but I agree to wait if we are introducing more build systems.

File: src/compile_commands.cc (hunk 1 above)
Comment (eidheim) [inline src/compile_commands.cc]: The function gets more complicated if copy_if is to be used:
```c++
std::vector<std::string> CompileCommands::Command::paramter_values(const std::string &parameter_name) const {
  std::vector<std::string> parameter_values(parameters);
  
  bool found_argument=false;
  auto it=std::copy_if(parameters.begin(), parameters.end(), parameter_values.begin(),
               [&parameter_name, &found_argument](const std::string &parameter){
    if(found_argument) {
      found_argument=false;
      return true;
    }
    else if(parameter==parameter_name) {
      found_argument=true;
      return false;
    }
    return false;
  });
  parameter_values.resize(std::distance(parameter_values.begin(), it));
  
  return parameter_values;
}
```

File: src/meson.cc (hunk 4 above)
Comment (eidheim) [inline src/meson.cc]: Will clean up this as well.

File: src/compile_commands.cc (hunk 1 above)
Comment (zalox) [inline src/compile_commands.cc]: std::copy_if(parameters.begin(),paramerters.end(),parameter_values.end(),[&parameter_name]const std::string &param(){
  return parameter_name == param;
}));

Comment (eidheim) [inline src/compile_commands.cc]: This is not the same. The idea is to return for instance `some_file` after `-o some_file` when the `parameter_name` is `-o`.

Comment (zalox) [inline src/compile_commands.cc]: Ohh, sorry I missed the else if, but at least this made me notice the miss spelt `paramter_name`.

Comment (eidheim) [inline src/compile_commands.cc]: I fixed the misspelling yesterday:)

**Comment by eidheim** [comment]
@zalox Are you happy with the changes? Can we merge this PR?

**Comment by eidheim** [comment]
@zalox the spelling errors should be fixed now

**Comment by zalox** [comment]
Looks good, sorry about my late responses, been flying all over Europe!

**Pull request merged**